Reduce macro usage for lints

nnethercote opened the pull request:
r? @cjgillot

nnethercote committed:
Eliminate four unnecessary lint macros.

The lint definitions use macros heavily. This commit merges some of them
that are split unnecessarily. I find the reduced indirection makes it
easier to imagine what the generated code will look like.

nnethercote committed:
Remove `-Zno-interleave-lints`.

Because it complicates lint implementation greatly.

nnethercote committed:
Remove six macros relating to lint definitions.

These were enabled by the removal of `-Zno-interleave-lints`.

nnethercote commented:
Commit 2 removes the `-Zno-interleave-lints` flag. I'm not sure if this is a good idea, but I was curious about how much simplification could occur with its removal. The answer is: quite a bit.

Nilstrieb commented:
The option was added in #57726 to enable better benchmarking of lints.

nnethercote commented:
Sure... the question is whether it's useful. Is anyone benchmarking lints with this option?

jyn514 commented:
r=me for the first commit; if you want to keep the others I think you should make an MCP for removing the flag.

Btw, IMO it would be nice to remove the `'tcx` macro argument and just hard-code tcx everywhere it's used, I don't see the advantage of splitting the two.

cjgillot commented on compiler/rustc_lint/src/late.rs:
```diff
@@ -417,33 +405,11 @@ fn late_lint_crate<'tcx, T: LateLintPass<'tcx>>(tcx: TyCtxt<'tcx>, builtin_lints
+        late_lint_pass_crate(tcx, LateLintPassObjects { lints: &mut passes[..] });
     }
+
+    late_lint_pass_crate(tcx, builtin_lints);
```
`builtin_lints` is always `BuiltinCombinedLateLintPass::new()`.

nnethercote replied:
There is another `check_crate` call in rustdoc, [here](https://github.com/rust-lang/rust/blob/c97b539e408ea353f4fde2f9251d598291fec421/src/librustdoc/core.rs#L344), so `builtin_lints` is required.

cjgillot commented on compiler/rustc_lint/src/late.rs:
```diff
@@ -370,11 +363,6 @@ pub fn late_lint_mod<'tcx, T: LateLintPass<'tcx>>(
     module_def_id: LocalDefId,
     builtin_lints: T,
```
This is always `BuiltinCombinedModuleLateLintPass::new()`.

nnethercote replied:
ditto

cjgillot commented on compiler/rustc_lint/src/late.rs:
```diff
@@ -417,33 +405,11 @@ fn late_lint_crate<'tcx, T: LateLintPass<'tcx>>(tcx: TyCtxt<'tcx>, builtin_lints
     }
+
+    late_lint_pass_crate(tcx, builtin_lints);
 }
```
Can this function be merged with `late_lint_mod`?

nnethercote replied:
I'm not sure which function is "this function", because it looks to me like this comment is on `late_lint_mod`. But as mentioned elsewhere, based on your other suggestions I was able to merge `late_lint_mod_pass` into `late_lint_mod`, and also `late_lint_pass_crate` into `late_lint_pass`.

It's not possible to merge `late_lint_mod` and `late_lint_crate`, though.

cjgillot commented on compiler/rustc_lint/src/late.rs:
```diff
@@ -417,33 +405,11 @@ fn late_lint_crate<'tcx, T: LateLintPass<'tcx>>(tcx: TyCtxt<'tcx>, builtin_lints
-                });
-        }
+    if !passes.is_empty() {
+        late_lint_pass_crate(tcx, LateLintPassObjects { lints: &mut passes[..] });
```
Can we merge the two traversals by adding the `builtin_lints` to `LateLintPassObjects`?

nnethercote replied:
Yes! That will work in both `late_lint_crate` and `late_lint_mod`. It will then let `late_lint_pass_crate` and `late_lint_mod_pass` be inlined and removed, because they'll both have a single call site.

nnethercote replied:
The same thing can be done for early lints.

cjgillot commented on compiler/rustc_lint/src/early.rs:
```diff
@@ -409,43 +401,26 @@ pub fn check_ast_node<'a>(
             registered_tools,
             buffered,
-            builtin_lints,
+            EarlyLintPassObjects { lints: &mut passes[..] },
```
Can we merge the two traversals by adding the `builtin_lints` to `EarlyLintPassObjects`?

nnethercote replied:
ditto

cjgillot reviewed (approved):
I'm not really concerned about `-Zno-interleave-lints`.
It has quite a significant complexity cost for a profiling option, and I'm not sure anybody used it besides who introduced it.
`tracing` is proably enough for timings, and profiling tools do not need this.

nnethercote committed:
Merge `builtins` into `LateLintPassObjects`.

This avoids calling the `late_lint_{mod_pass,pass_crate}` twice.

nnethercote committed:
Inline and remove `late_lint_mod_pass`.

It has a single call site.

nnethercote committed:
Inline and remove `late_lint_pass_crate`.

It has a single call site.

nnethercote committed:
Merge `builtins` into `EarlyLintPassObjects`.

This avoids calling `early_lint_node` twice.

Note: one `early_lint_node` call had `!pre_expansion` for the second
argument and the other had `false`. The new single call just has
`!pre_expansion`. This results in a reduction of duplicate error
messages in some `ui-fulldeps` tests. The order of some `ui-fulldeps`
output also changes, but that doesn't matter.

nnethercote committed:
Inline and remove `early_lint_node`.

It has a single call site.

nnethercote commented:
I have added five new commits that put `builtins` in with other passes, which then allows some functions to be inlined and removed.

I feel like this doesn't need an MCP. There was no MCP for #102725 which was similar to this -- removing an obscure unstable flag that is only used for compiling rustc itself.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3589203212/jobs/6041396137) [(plain)](https://github.com/rust-lang/rust/commit/aee176d9e6cca88319594c3d0dc249314298a2d0/checks/9812526939/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_const_eval v0.0.0 (/checkout/compiler/rustc_const_eval)
    Checking rustc_traits v0.0.0 (/checkout/compiler/rustc_traits)
    Checking rustc_ty_utils v0.0.0 (/checkout/compiler/rustc_ty_utils)
    Checking rustc_mir_build v0.0.0 (/checkout/compiler/rustc_mir_build)
error[E0277]: `impl EarlyLintPass + 'static` cannot be sent between threads safely
    |
378 |     passes.push(Box::new(builtin_lints));
378 |     passes.push(Box::new(builtin_lints));
    |                 ^^^^^^^^^^^^^^^^^^^^^^^ `impl EarlyLintPass + 'static` cannot be sent between threads safely
    |
    = note: required for the cast from `impl EarlyLintPass + 'static` to the object type `dyn EarlyLintPass + Send`
    |
    |
372 |     builtin_lints: impl EarlyLintPass + 'static + std::marker::Send,


error[E0277]: `T` cannot be sent between threads safely
    |
348 |     passes.push(Box::new(builtin_lints));
348 |     passes.push(Box::new(builtin_lints));
    |                 ^^^^^^^^^^^^^^^^^^^^^^^ `T` cannot be sent between threads safely
    |
    = note: required for the cast from `T` to the object type `dyn LateLintPass<'_> + Send`
    |
    |
329 | pub(super) fn late_lint_mod<'tcx, T: LateLintPass<'tcx> + 'tcx + std::marker::Send>(


error[E0277]: `T` cannot be sent between threads safely
    |
379 |     passes.push(Box::new(builtin_lints));
379 |     passes.push(Box::new(builtin_lints));
    |                 ^^^^^^^^^^^^^^^^^^^^^^^ `T` cannot be sent between threads safely
    |
    = note: required for the cast from `T` to the object type `dyn LateLintPass<'_> + Send`
    |
    |
364 | fn late_lint_crate<'tcx, T: LateLintPass<'tcx> + 'tcx + std::marker::Send>(tcx: TyCtxt<'tcx>, builtin_lints: T) {

For more information about this error, try `rustc --explain E0277`.
error: could not compile `rustc_lint` due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
```

</details>

nnethercote commented:
This is odd, I don't get this error when I compile locally:
```
error[E0277]: `impl EarlyLintPass + 'static` cannot be sent between threads safely
    |
378 |     passes.push(Box::new(builtin_lints));
378 |     passes.push(Box::new(builtin_lints));
    |                 ^^^^^^^^^^^^^^^^^^^^^^^ `impl EarlyLintPass + 'static` cannot be sent between threads safely
    |
    = note: required for the cast from `impl EarlyLintPass + 'static` to the object type `dyn EarlyLintPass + Send`
    |
    |
372 |     builtin_lints: impl EarlyLintPass + 'static + std::marker::Send,
```

nnethercote committed:
Remove some unnecessary `Send` bounds.

Required to get the parallel compiler building again.

nnethercote commented:
Looks like the failure was in a `parallel-compiler = true` build. I've added another commit to fix that.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3589576660/jobs/6042144568) [(plain)](https://github.com/rust-lang/rust/commit/e711bf74bdd2c74cc776620db842ced12f75f01d/checks/9813455732/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Copying stage2 rustc from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Check compiletest suite=ui-fulldeps mode=ui (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 71 tests
Some tests failed in compiletest suite=ui-fulldeps mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
.......................i.............F...F.F.FF..F......F.......F......

---- [ui] src/test/ui-fulldeps/lint-plugin-deny-attr.rs stdout ----
diff of stderr:


- warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
-   --> $DIR/lint-plugin-deny-attr.rs:5:1
-    |
- LL | #![plugin(lint_plugin_test)]
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
-    = note: `#[warn(deprecated)]` on by default
- 
- 
9 error: item is named 'lintme'
11    |

17    |
18 LL | #![deny(test_lint)]
18 LL | #![deny(test_lint)]
19    |         ^^^^^^^^^
+ 
+ warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
+    |
+    |
+ LL | #![plugin(lint_plugin_test)]
+    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
+    = note: `#[warn(deprecated)]` on by default
20 
21 error: aborting due to previous error; 1 warning emitted
22 
---
To only update this specific test, also pass `--test-args lint-plugin-deny-attr.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui-fulldeps/lint-plugin-deny-attr.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-plugin-deny-attr" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-plugin-deny-attr/auxiliary"
stdout: none
--- stderr -------------------------------
error: item is named 'lintme'
   |
   |
LL | fn lintme() { } //~ ERROR item is named 'lintme'
   |
note: the lint level is defined here
  --> /checkout/src/test/ui-fulldeps/lint-plugin-deny-attr.rs:7:9
   |
   |
LL | #![deny(test_lint)]
   |         ^^^^^^^^^

warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
   |
   |
LL | #![plugin(lint_plugin_test)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
   = note: `#[warn(deprecated)]` on by default

error: aborting due to previous error; 1 warning emitted
------------------------------------------
------------------------------------------


---- [ui] src/test/ui-fulldeps/lint-plugin-cmdline-load.rs stdout ----
diff of stderr:

- warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
-   --> <crate attribute>:1:1
-    |
- LL | plugin(lint_plugin_test)
-    | ^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
-    = note: `#[warn(deprecated)]` on by default
- 
- 
9 warning: item is named 'lintme'
11    |

13    | ^^^^^^^^^^^^^^^
14    |
14    |
15    = note: `#[warn(test_lint)]` on by default
+ 
+ warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
+   --> <crate attribute>:1:1
+    |
+ LL | plugin(lint_plugin_test)
+    | ^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
+    = note: `#[warn(deprecated)]` on by default
16 
17 warning: 2 warnings emitted
18 
---
To only update this specific test, also pass `--test-args lint-plugin-cmdline-load.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui-fulldeps/lint-plugin-cmdline-load.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-plugin-cmdline-load" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-plugin-cmdline-load/auxiliary" "-Z" "crate-attr=plugin(lint_plugin_test)"
stdout: none
--- stderr -------------------------------
warning: item is named 'lintme'
   |
   |
LL | fn lintme() { } //~ WARNING item is named 'lintme'
   |
   |
   = note: `#[warn(test_lint)]` on by default

warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
  --> <crate attribute>:1:1
   |
LL | plugin(lint_plugin_test)
   | ^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
   = note: `#[warn(deprecated)]` on by default

warning: 2 warnings emitted
------------------------------------------
------------------------------------------


---- [ui] src/test/ui-fulldeps/lint-plugin-forbid-cmdline.rs stdout ----
diff of stderr:

6    |
7    = note: `forbid` lint level was set on command line
8 
- error[E0453]: allow(test_lint) incompatible with previous forbid
-   --> $DIR/lint-plugin-forbid-cmdline.rs:10:9
-    |
- LL | #[allow(test_lint)]
-    |         ^^^^^^^^^ overruled by previous forbid
-    |
-    = note: `forbid` lint level was set on command line
- 
- warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
-   --> $DIR/lint-plugin-forbid-cmdline.rs:6:1
-    |
- LL | #![plugin(lint_plugin_test)]
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
-    = note: `#[warn(deprecated)]` on by default
- 
- 
25 error: item is named 'lintme'
27    |

38    |
38    |
39    = note: `forbid` lint level was set on command line
- error: aborting due to 4 previous errors; 1 warning emitted
- error: aborting due to 4 previous errors; 1 warning emitted
+ warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
+    |
+    |
+ LL | #![plugin(lint_plugin_test)]
+    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
+    = note: `#[warn(deprecated)]` on by default
+ 
+ error: aborting due to 3 previous errors; 1 warning emitted
42 
---
To only update this specific test, also pass `--test-args lint-plugin-forbid-cmdline.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui-fulldeps/lint-plugin-forbid-cmdline.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-plugin-forbid-cmdline" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-plugin-forbid-cmdline/auxiliary" "-F" "test-lint"
stdout: none
--- stderr -------------------------------
error[E0453]: allow(test_lint) incompatible with previous forbid
   |
   |
LL | #[allow(test_lint)] //~ ERROR allow(test_lint) incompatible
   |         ^^^^^^^^^ overruled by previous forbid
   |
   = note: `forbid` lint level was set on command line

error: item is named 'lintme'
   |
   |
LL | fn lintme() { } //~ ERROR item is named 'lintme'
   |
   |
   = note: requested on the command line with `-F test-lint`

error[E0453]: allow(test_lint) incompatible with previous forbid
   |
   |
LL | #[allow(test_lint)] //~ ERROR allow(test_lint) incompatible
   |         ^^^^^^^^^ overruled by previous forbid
   |
   = note: `forbid` lint level was set on command line

warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
   |
   |
LL | #![plugin(lint_plugin_test)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
   = note: `#[warn(deprecated)]` on by default

error: aborting due to 3 previous errors; 1 warning emitted


For more information about this error, try `rustc --explain E0453`.
------------------------------------------


---- [ui] src/test/ui-fulldeps/lint-plugin-deny-cmdline.rs stdout ----
diff of stderr:

- warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
-   --> $DIR/lint-plugin-deny-cmdline.rs:6:1
-    |
- LL | #![plugin(lint_plugin_test)]
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
-    = note: `#[warn(deprecated)]` on by default
- 
- 
9 error: item is named 'lintme'
11    |

13    | ^^^^^^^^^^^^^^^
14    |
14    |
15    = note: requested on the command line with `-D test-lint`
+ 
+ warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
+    |
+    |
+ LL | #![plugin(lint_plugin_test)]
+    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
+    = note: `#[warn(deprecated)]` on by default
16 
17 error: aborting due to previous error; 1 warning emitted
18 
---
To only update this specific test, also pass `--test-args lint-plugin-deny-cmdline.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui-fulldeps/lint-plugin-deny-cmdline.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-plugin-deny-cmdline" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-plugin-deny-cmdline/auxiliary" "-D" "test-lint"
stdout: none
--- stderr -------------------------------
error: item is named 'lintme'
   |
   |
LL | fn lintme() { } //~ ERROR item is named 'lintme'
   |
   |
   = note: requested on the command line with `-D test-lint`

warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
   |
   |
LL | #![plugin(lint_plugin_test)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
   = note: `#[warn(deprecated)]` on by default

error: aborting due to previous error; 1 warning emitted
------------------------------------------
------------------------------------------


---- [ui] src/test/ui-fulldeps/lint-plugin-forbid-attrs.rs stdout ----
diff of stderr:

7 LL | #[allow(test_lint)]
8    |         ^^^^^^^^^ overruled by previous forbid
9 
- error[E0453]: allow(test_lint) incompatible with previous forbid
-   --> $DIR/lint-plugin-forbid-attrs.rs:11:9
-    |
- LL | #![forbid(test_lint)]
-    |           --------- `forbid` level set here
- ...
- LL | #[allow(test_lint)]
-    |         ^^^^^^^^^ overruled by previous forbid
- 
- warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
-   --> $DIR/lint-plugin-forbid-attrs.rs:5:1
-    |
- LL | #![plugin(lint_plugin_test)]
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
-    = note: `#[warn(deprecated)]` on by default
- 
- 
27 error: item is named 'lintme'
29    |


45 LL | #[allow(test_lint)]
46    |         ^^^^^^^^^ overruled by previous forbid
- error: aborting due to 4 previous errors; 1 warning emitted
- error: aborting due to 4 previous errors; 1 warning emitted
+ warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
+    |
+    |
+ LL | #![plugin(lint_plugin_test)]
+    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
+    = note: `#[warn(deprecated)]` on by default
+ 
+ error: aborting due to 3 previous errors; 1 warning emitted
49 
---
To only update this specific test, also pass `--test-args lint-plugin-forbid-attrs.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui-fulldeps/lint-plugin-forbid-attrs.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-plugin-forbid-attrs" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-plugin-forbid-attrs/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0453]: allow(test_lint) incompatible with previous forbid
   |
   |
LL | #![forbid(test_lint)]
   |           --------- `forbid` level set here
...
LL | #[allow(test_lint)]
   |         ^^^^^^^^^ overruled by previous forbid

error: item is named 'lintme'
   |
   |
LL | fn lintme() {} //~ ERROR item is named 'lintme'
   |
note: the lint level is defined here
  --> /checkout/src/test/ui-fulldeps/lint-plugin-forbid-attrs.rs:7:11
   |
   |
LL | #![forbid(test_lint)]


error[E0453]: allow(test_lint) incompatible with previous forbid
   |
   |
LL | #![forbid(test_lint)]
   |           --------- `forbid` level set here
...
LL | #[allow(test_lint)]
   |         ^^^^^^^^^ overruled by previous forbid

warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
   |
   |
LL | #![plugin(lint_plugin_test)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
   = note: `#[warn(deprecated)]` on by default

error: aborting due to 3 previous errors; 1 warning emitted


For more information about this error, try `rustc --explain E0453`.
------------------------------------------


---- [ui] src/test/ui-fulldeps/lint-plugin.rs stdout ----
diff of stderr:

- warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
-   --> $DIR/lint-plugin.rs:5:1
-    |
- LL | #![plugin(lint_plugin_test)]
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
-    = note: `#[warn(deprecated)]` on by default
- 
- 
9 warning: item is named 'lintme'
11    |

13    | ^^^^^^^^^^^^^^^
14    |
14    |
15    = note: `#[warn(test_lint)]` on by default
+ 
+ warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
+    |
+    |
+ LL | #![plugin(lint_plugin_test)]
+    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
+    = note: `#[warn(deprecated)]` on by default
16 
17 warning: 2 warnings emitted
18 
---
To only update this specific test, also pass `--test-args lint-plugin.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui-fulldeps/lint-plugin.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "-O" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-plugin/a" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-plugin/auxiliary"
stdout: none
--- stderr -------------------------------
warning: item is named 'lintme'
   |
   |
LL | fn lintme() { } //~ WARNING item is named 'lintme'
   |
   |
   = note: `#[warn(test_lint)]` on by default

warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
   |
   |
LL | #![plugin(lint_plugin_test)] //~ WARNING use of deprecated attribute
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
   = note: `#[warn(deprecated)]` on by default

warning: 2 warnings emitted
------------------------------------------
------------------------------------------


---- [ui] src/test/ui-fulldeps/lint-tool-cmdline-allow.rs stdout ----
diff of stderr:

6    |
7    = note: requested on the command line with `-A test_lint`
8 
+ warning: item is named 'lintme'
+    |
+    |
+ LL | fn lintme() {}
+    |
+    |
+    = note: `#[warn(clippy::test_lint)]` on by default
+ 
9 warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
11    |

18    |
18    |
19    = note: requested on the command line with `-A test_lint`
20 
- warning: item is named 'lintme'
-   --> $DIR/lint-tool-cmdline-allow.rs:9:1
-    |
- LL | fn lintme() {}
-    |
-    |
-    = note: `#[warn(clippy::test_lint)]` on by default
- 
- warning: lint name `test_lint` is deprecated and does not have an effect anymore. Use: clippy::test_lint
-    |
-    = note: requested on the command line with `-A test_lint`
- warning: 6 warnings emitted
+ warning: 5 warnings emitted
34 
35 
---
To only update this specific test, also pass `--test-args lint-tool-cmdline-allow.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui-fulldeps/lint-tool-cmdline-allow.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-tool-cmdline-allow" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-tool-cmdline-allow/auxiliary" "-A" "test-lint"
stdout: none
--- stderr -------------------------------
warning: lint name `test_lint` is deprecated and does not have an effect anymore. Use: clippy::test_lint
   |
   = note: requested on the command line with `-A test_lint`

warning: lint name `test_lint` is deprecated and does not have an effect anymore. Use: clippy::test_lint
   |
   = note: requested on the command line with `-A test_lint`

warning: item is named 'lintme'
   |
   |
LL | fn lintme() {}
   |
   |
   = note: `#[warn(clippy::test_lint)]` on by default

warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
   |
   |
LL | #![plugin(lint_tool_test)] //~ WARNING compiler plugins are deprecated
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
   = note: `#[warn(deprecated)]` on by default


warning: lint name `test_lint` is deprecated and does not have an effect anymore. Use: clippy::test_lint
   |
   = note: requested on the command line with `-A test_lint`
warning: 5 warnings emitted
------------------------------------------



---- [ui] src/test/ui-fulldeps/lint-tool-test.rs stdout ----
diff of stderr:

30 LL | #![deny(clippy_group)]
31    |         ^^^^^^^^^^^^ help: change it to: `clippy::group`
32 
- warning: lint name `test_group` is deprecated and may not have an effect in the future.
-   --> $DIR/lint-tool-test.rs:31:9
-    |
- LL | #[allow(test_group)]
-    |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
- warning: unknown lint: `this_lint_does_not_exist`
-   --> $DIR/lint-tool-test.rs:36:8
-    |
-    |
- LL | #[deny(this_lint_does_not_exist)]
-    |
-    = note: `#[warn(unknown_lints)]` on by default
- 
- 
- warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
-   --> $DIR/lint-tool-test.rs:6:1
-    |
- LL | #![plugin(lint_tool_test)]
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
-    = note: `#[warn(deprecated)]` on by default
- 
- warning: lint name `test_lint` is deprecated and may not have an effect in the future.
-   --> $DIR/lint-tool-test.rs:9:23
-   --> $DIR/lint-tool-test.rs:9:23
-    |
- LL | #![cfg_attr(foo, warn(test_lint))]
-    |                       ^^^^^^^^^ help: change it to: `clippy::test_lint`
- 
- warning: lint name `clippy_group` is deprecated and may not have an effect in the future.
-   --> $DIR/lint-tool-test.rs:14:9
-    |
- LL | #![deny(clippy_group)]
-    |         ^^^^^^^^^^^^ help: change it to: `clippy::group`
- 
67 error: item is named 'lintme'
69    |


91 LL | #[allow(test_group)]
92    |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
+ warning: unknown lint: `this_lint_does_not_exist`
+   --> $DIR/lint-tool-test.rs:36:8
+    |
+    |
+ LL | #[deny(this_lint_does_not_exist)]
+    |
+    = note: `#[warn(unknown_lints)]` on by default
+ 
+ 
+ warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
+    |
+    |
+ LL | #![plugin(lint_tool_test)]
+    | ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
+    = note: `#[warn(deprecated)]` on by default
+ 
94 warning: lint name `test_lint` is deprecated and may not have an effect in the future.
95   --> $DIR/lint-tool-test.rs:9:23
95   --> $DIR/lint-tool-test.rs:9:23
96    |

109 LL | #[allow(test_group)]
110    |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
- error: aborting due to 2 previous errors; 14 warnings emitted
+ error: aborting due to 2 previous errors; 11 warnings emitted
113 
114 
---
To only update this specific test, also pass `--test-args lint-tool-test.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui-fulldeps/lint-tool-test.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-tool-test" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-tool-test/auxiliary" "--cfg" "foo"
stdout: none
--- stderr -------------------------------
warning: lint name `test_lint` is deprecated and may not have an effect in the future.
   |
   |
LL | #![cfg_attr(foo, warn(test_lint))]
   |                       ^^^^^^^^^ help: change it to: `clippy::test_lint`
   = note: `#[warn(renamed_and_removed_lints)]` on by default


warning: lint name `clippy_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #![deny(clippy_group)]
   |         ^^^^^^^^^^^^ help: change it to: `clippy::group`

warning: lint name `test_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #[allow(test_group)]
   |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
warning: lint name `test_lint` is deprecated and may not have an effect in the future.
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:9:23
   |
   |
LL | #![cfg_attr(foo, warn(test_lint))]
   |                       ^^^^^^^^^ help: change it to: `clippy::test_lint`

warning: lint name `clippy_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #![deny(clippy_group)]
   |         ^^^^^^^^^^^^ help: change it to: `clippy::group`

error: item is named 'lintme'
   |
   |
LL | fn lintme() { } //~ ERROR item is named 'lintme'
   |
note: the lint level is defined here
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:14:9
   |
   |
LL | #![deny(clippy_group)]
   |         ^^^^^^^^^^^^
   = note: `#[deny(clippy::test_lint)]` implied by `#[deny(clippy::group)]`

error: item is named 'lintmetoo'
   |
   |
LL |     fn lintmetoo() { } //~ ERROR item is named 'lintmetoo'
   |
   |
   = note: `#[deny(clippy::test_group)]` implied by `#[deny(clippy::group)]`

warning: lint name `test_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #[allow(test_group)]
   |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
warning: unknown lint: `this_lint_does_not_exist`
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:36:8
   |
   |
LL | #[deny(this_lint_does_not_exist)] //~ WARNING unknown lint: `this_lint_does_not_exist`
   |
   = note: `#[warn(unknown_lints)]` on by default


warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
   |
   |
LL | #![plugin(lint_tool_test)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
   = note: `#[warn(deprecated)]` on by default

warning: lint name `test_lint` is deprecated and may not have an effect in the future.
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:9:23
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:9:23
   |
LL | #![cfg_attr(foo, warn(test_lint))]
   |                       ^^^^^^^^^ help: change it to: `clippy::test_lint`

warning: lint name `clippy_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #![deny(clippy_group)]
   |         ^^^^^^^^^^^^ help: change it to: `clippy::group`

warning: lint name `test_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #[allow(test_group)]
   |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
error: aborting due to 2 previous errors; 11 warnings emitted
------------------------------------------


```

</details>

nnethercote commented:
Some tests that only run with `--stage=2` were failing. Turns out we can't merge the early lints into a single pass, presumably because the second argument to `early_lint_node` differs in the current two call sites.

cjgillot commented:
The second argument to `early_lint_node` just says whether to warn unknown lints. It should be true exactly once in the runtime, so keeping `!pre_expansion` should be it.

nnethercote commented:
Well, I got those `ui-fulldeps` failures with the early lint changes.

nnethercote commented:
I tried the early lint change again. Most of the `ui-fulldeps` failures are resolvable, because the change causes some duplicate warnings to be removed, which is fine.

But there is one remaining bizarre failure:
```
---- [ui] src/test/ui-fulldeps/mod_dir_path_canonicalized.rs stdout ----

error: test run failed!
status: exit status: 101
command: "/home/njn/dev/rust1/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/mod_dir_path_canonicalized/a"
stdout: none
--- stderr -------------------------------
thread 'main' panicked at 'WorkerLocal can only be used on the thread pool it was created on', /home/njn/.cargo/registry/src/github.com-1ecc6299db9ec823/rustc-rayon-core-0.4.1/src/worker_local.rs:49:17
stack backtrace:
   0: std::panicking::begin_panic::<&str>
   1: rustc_ast::attr::mk_attr_from_item
   2: <rustc_parse::parser::Parser>::collect_tokens_trailing_token::<rustc_ast::ast::Attribute, <rustc_parse::parser::Parser>::collect_tokens_no_attrs<rustc_ast::ast::Attribute, <rustc_parse::parser::Parser>::parse_attribute::{closure#0}>::{closure#0}>
   3: <rustc_parse::parser::Parser>::parse_inner_attributes
   4: <rustc_parse::parser::Parser>::parse_mod
   5: <rustc_parse::parser::Parser>::parse_crate_mod
   6: mod_dir_path_canonicalized::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
------------------------------------------

failures:
    [ui] src/test/ui-fulldeps/mod_dir_path_canonicalized.rs
```

nnethercote commented:
Oh, I was accidentally using a parallel compiler build, sigh.

nnethercote commented:
Ok, the commits to merge `early_lint_node` are now working and ready for review.

cjgillot commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 406dace6f296c818438620b558b1add62b6e55a0 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 406dace6f296c818438620b558b1add62b6e55a0 -->
<!-- homu: {"type":"Approved","sha":"406dace6f296c818438620b558b1add62b6e55a0","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 406dace6f296c818438620b558b1add62b6e55a0 with merge e960b5e7749e95c6a6b2fdec7250a48105664efb...
<!-- homu: {"type":"BuildStarted","head_sha":"406dace6f296c818438620b558b1add62b6e55a0","merge_sha":"e960b5e7749e95c6a6b2fdec7250a48105664efb"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3602984607/jobs/6073253636)
Approved by: cjgillot
Pushing e960b5e7749e95c6a6b2fdec7250a48105664efb to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3602984607/jobs/6073253636"},"merge_sha":"e960b5e7749e95c6a6b2fdec7250a48105664efb"} -->

rust-timer commented:
Finished benchmarking commit (e960b5e7749e95c6a6b2fdec7250a48105664efb): [comparison URL](https://perf.rust-lang.org/compare.html?start=cef44f53034eac46be3a0e3eec7b2b3d4ef5140b&end=e960b5e7749e95c6a6b2fdec7250a48105664efb&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=cef44f53034eac46be3a0e3eec7b2b3d4ef5140b&end=e960b5e7749e95c6a6b2fdec7250a48105664efb&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5% | [0.2%, 1.4%] | 116   |
| Regressions ❌ <br /> (secondary)  | 0.9% | [0.1%, 2.3%] | 83    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.5% | [0.2%, 1.4%] | 116   |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=cef44f53034eac46be3a0e3eec7b2b3d4ef5140b&end=e960b5e7749e95c6a6b2fdec7250a48105664efb&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.2%  | [3.2%, 3.2%]   | 1     |
| Improvements ✅ <br /> (primary)   | -2.7% | [-2.7%, -2.7%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.5% | [-3.6%, -1.4%] | 2     |
| All ❌✅ (primary)                 | -2.7% | [-2.7%, -2.7%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=cef44f53034eac46be3a0e3eec7b2b3d4ef5140b&end=e960b5e7749e95c6a6b2fdec7250a48105664efb&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

nnethercote commented:
Ugh, that's unexpected. Lots of instruction count regressions, but cycles and wall times didn't change. I'll take a look on Monday.

nnethercote commented:
These were the two commits that caused the regressions:
```
890c5ead204 Merge `builtins` into `LateLintPassObjects`.
a9b02af62b0 Merge `builtins` into `EarlyLintPassObjects`.
```
The cachegrind diffs indicate that it's not because any extra work was being done. It just appears to be due to minor changes in code generation, mostly around `enter_lint_attrs`/`exit_lint_attrs`.

nnethercote commented:
#105291 improved things slightly, though less than I had hoped based on local measurements.

I have another idea, an idea which I had before I wrote this PR, and and idea for which this PR is a prerequisite. Currently when processing a node such as an expression, we call `check_expr` for every lint pass, even those which have an empty `check_expr` methods. And because we use dynamic dispatch for this, the cost of a call to an empty method isn't zero. And we have over 30 `check_*` methods, and a couple of dozen lint passes, and most of them only define a small number of the `check_*` methods. Which means that there are a lot of calls to empty `check_*` methods. It may be possible to avoid these.

nnethercote commented:
> I have another idea, an idea which I had before I wrote this PR, and and idea for which this PR is a prerequisite. Currently when processing a node such as an expression, we call `check_expr` for every lint pass, even those which have an empty `check_expr` methods. And because we use dynamic dispatch for this, the cost of a call to an empty method isn't zero. And we have over 30 `check_*` methods, and a couple of dozen lint passes, and most of them only define a small number of the `check_*` methods. Which means that there are a lot of calls to empty `check_*` methods. It may be possible to avoid these.

I was wrong about this. Well, it's relevant for clippy, but for rustc all the builtin lints get merged into a small number of "combined" lints via a macro-based mechanism, which lets the empty `check_*` methods be inlined and thus disappear.

I also tried tweaking some inlining in #105416. It helped locally, but didn't make much difference for a CI perf run.

nnethercote commented:
#105485 should fix the perf regression properly, by reverting part of this PR's changes.

nnethercote commented:
@rustbot label: +perf-regression-triaged
